### PR TITLE
loop over value keys to check wich value need to be updated

### PR DESCRIPTION
**Pull request by christophe-g**
Should resolve https://github.com/firebase/polymerfire/issues/88 and https://github.com/firebase/polymerfire/issues/56.

### CODE REVIEW

File: firebase-document.html [hunk 1]
```diff
@@ -186,11 +187,23 @@
...
-      if (!this.new) {
+      if (!this.isNew) {
         this.async(function() {
           this.syncToMemory(function() {
             this._log('Updating data from Firebase value:', value);
-            this.set('data', value);
+            
+            // set the value if this.data does not exist or value is primitive or if firebase value obj contain less keys than this.data (https://github.com/Polymer/polymer/issues/2565)
+            if (this.__firstSync || !this.data || typeof value !== 'object' || ( Object.keys(value).length <  Object.keys(this.data).length)) {
+              delete this.__firstSync;
```
Comment (mbleigh) [inline firebase-document.html]: Prefer `this.__firstSync = false`

File: firebase-document.html [hunk 2]
```diff
@@ -186,11 +187,23 @@
...
+              return this.set('data', value);
+            }
+
+            // now, we loop over keys
+            for (var prop in value) {
+              if(value[prop] !== this.data[prop]) {
```
Comment (mbleigh) [inline firebase-document.html]: This will always be true for sub-objects, no?

**Comment by christophe-g** [comment]
@mbleigh thanks for your comments, truly appreciated. 

For very large JSON blob, I would tend to think that other part of data processing will weight on performance way more than Object.key() (I am thinking of snapshot.val() in particular - which is process intensive and invoked by firebase-document on every firebase value update). 

Re. nested Object properties - yes they will still trigger on every update. I was reluctant to recurse down the entire tree because the problem this PR is trying to fix is to avoid observers to be fired every time a an update is made on a bound Object. A construct like `_observe(data.myProp)` is pretty common, whereas `observe(data.myProp.mySubProp)` is a bit less likely.

File: firebase-document.html (hunk 2 above)
Comment (christophe-g) [inline firebase-document.html]: yes it will - see my comment below for a short rationale

File: firebase-document.html (hunk 1 above)
Comment (christophe-g) [inline firebase-document.html]: o.k. - will update that

File: firebase-document.html [hunk 3]
```diff
@@ -186,11 +187,27 @@
         value = this.zeroValue;
       }
 
-      if (!this.new) {
+      if (!this.isNew) {
```
Comment (chuckh) [inline firebase-document.html]: Should this be  `if (!this.isNew()) {`?

Comment (christophe-g) [inline firebase-document.html]: isNew is defined via its getter
```js
    get isNew() {
      return this.disabled || !this.__pathReady(this.path);
    },
```

File: firebase-database-behavior.html [hunk 4]
```diff
@@ -90,9 +90,10 @@
       },
 
       __pathChanged: function(path, oldPath) {
-        if (oldPath != null && !this.disabled && this.__pathReady(path)) {
+        if (!this.disabled && !this.valueIsEmpty(this.data)) {
```
Comment (christophe-g) [inline firebase-database-behavior.html]: we also want to handle the case where `path` is not a valid one,  but `oldPath` was.

**Comment by christophe-g** [comment]
Please note that if you want to and as a temporary measure, you can use my #dev version combining pull request #177, #166 and #167 via 
```
bower install --save  PolymerEl/polymerfire#dev
```

**Pull request merged**